django/utils/functional.py

```diff
@@ -223,12 +224,40 @@ def keep_lazy(*resultclasses):
     def decorator(func):
         lazy_func = lazy(func, *resultclasses)
 
+        func_params = tuple(inspect.signature(func).parameters.items())
+        # Skip parameters which are *args, or **kwargs, which are intrinsically
+        # multiple arguments and must go through the slow path.
+        contains_var_parameters = bool({
+            k for k, v in func_params
+            if v.kind in (inspect.Parameter.VAR_POSITIONAL, inspect.Parameter.VAR_KEYWORD)
+        })
+        non_var_parameters = [
+            k for k, v in func_params
+            if v.kind not in (inspect.Parameter.VAR_POSITIONAL, inspect.Parameter.VAR_KEYWORD)
+        ]
+        func_first_param_name = None
+        if non_var_parameters:
+            func_first_param_name = non_var_parameters[0]
+
         @wraps(func)
-        def wrapper(*args, **kwargs):
+        def keep_lazy_single_argument_wrapper(*args, **kwargs):
+            if (args and isinstance(args[0], Promise)) or (
+                func_first_param_name in kwargs
+                and isinstance(kwargs[func_first_param_name], Promise)
+            ):
+                return lazy_func(*args, **kwargs)
+            return func(*args, **kwargs)
+
+        @wraps(func)
+        def keep_lazy_multiple_argument_wrapper(*args, **kwargs):
             if any(isinstance(arg, Promise) for arg in itertools.chain(args, kwargs.values())):
```

Comment:
If we `from itertools import chain` we can save the overhead of attribute access here.

Reply:
If we're going to touch the original implementation, it may be worth looking at the old version across the various cases (first arg non-promise/first arg promise, subsequent arg non-promise/subsequent arg promise), because IIRC that was _also_ a bit faster for something like `escape('<html>goes&here</html>')`, even if the new form is _perhaps_ more '_pythonic_' than:
```
            for arg in itertools.chain(args, kwargs.values()):
                if isinstance(arg, Promise):
                    break
            else:
                return func(*args, **kwargs)
            return lazy_func(*args, **kwargs)
```

I believe it was still slightly slower than making an optimized 1-arg variant (anecdotal datapoint of ... 1 laptop ;)), because as you note, attribute access, and neither `.chain()` nor `.values()` is _free_. But maybe just re-instituting that (if it still proves faster-than-current) would be a simpler middleground...